onnxruntime/core/providers/cpu/math/einsum_utils.cc

```diff
@@ -0,0 +1,724 @@
+  input_subscript_indices_.reserve(inputs_.size());
+
+  // We arbitrarily reserve space for 10 values as we don't expect to see any input with rank >10
+  // which would make num_subscript_indices_ > 10
+  subscript_indices_to_last_input_.reserve(10);
+  subscript_indices_to_dim_value_.reserve(10);
+
+  for (const auto& subscript : left_equation_split) {
+    const auto& shape = inputs_[input_index]->Shape();
+    const auto& dims = shape.GetDims();
+    size_t rank = dims.size();
+    size_t dim_counter = 0;
+
+    std::vector<int64_t> current_subscript_indices;
+    current_subscript_indices.reserve(rank);
+
+    // Temp variables to deal with "ellipsis" in the input
+    bool is_in_middle_of_ellipsis = false;
+    int64_t ellipsis_char_count = 0;
+
+    // Iterate through all subscript labels in the subscript
+    for (auto subscript_label : subscript) {
+      // Broadcasting based dims
+      if (subscript_label == '.') {
+        is_in_middle_of_ellipsis = true;
+        // Make sure there aren't more than 3 '.'s in the current subscript
+        ORT_ENFORCE(++ellipsis_char_count <= 3, "Found a '.' not part of an ellipsis in input: ", input_index);
+
+        // We have seen all 3 '.'s. We can safely process the ellipsis now.
+        if (ellipsis_char_count == 3) {
+          is_in_middle_of_ellipsis = false;
+
+          // Example for the following line of code
+          // Subscript "...ij" for an input of rank 6
+          // num_of_ellipsis_dims = 6 - 5 + 3 = 4
+          int64_t current_num_of_ellipsis_dims = rank - subscript.length() + 3;
+          ORT_ENFORCE(current_num_of_ellipsis_dims >= 0,
+                      "Einsum subscripts string contains too many subscript labels when compared to the rank of the input ",
+                      input_index);
+
+          // Theoretically, current_num_of_ellipsis_dims could be 0
+          // Example: For an input of rank 2 paired with a subscript "...ij"
+          if (current_num_of_ellipsis_dims != 0) {
+            // We have seen a ellipsis before - make sure ranks align as per the ONNX spec -
+            // "Ellipsis must indicate a fixed number of dimensions."
+            if (num_of_ellipsis_dims_ != 0) {
+              ORT_ENFORCE(num_of_ellipsis_dims_ == static_cast<size_t>(current_num_of_ellipsis_dims),
+                          "Ellipsis must indicate a fixed number of dimensions across all inputs");
+            } else {
+              num_of_ellipsis_dims_ = static_cast<size_t>(current_num_of_ellipsis_dims);
+            }
+
+            // We reserve '26' for broadcasted dims as we only allow 'a' - 'z' (0 - 25) for non-broadcasted dims
+            // We will assign appropriate indices (based o number of dimensions the ellipsis corresponds to) during broadcasting related post-processing
+            for (size_t i = 0; i < num_of_ellipsis_dims_; ++i) {
+              current_subscript_indices.push_back(26);
+            }
```

Comment:
Should this use num_of_letters? Could we use -1 instead of 26?

Reply:
Good idea. Used num_of_letters.